packages/gelato-core/hardware-adapters/rasp-gpio/rasp-gpio.ts

```
@@ -14,9 +13,15 @@ export class RaspGpio implements EventSensor<GpioEvent> {
     private gpioHardware = rpiGpio.promise
   ) {}
 
-  read(pin: number): GpioEvent {
-    const gpioEvent = {} as GpioEvent
-    return gpioEvent
+  async read(channel: string): Promise<GpioEvent> {
+    const event: GpioEvent = { value: null }
+    try {
+      let result = await this.gpioHardware.read(channel)
+      event.value = result ? GpioPinState.Low : GpioPinState.High
+    } catch (error) {
+      throw error
```

Any reason to catch and rethrow? One reason I can think of is we may want to throw a custom error.

---

Yes potentially. Should that function even attempt to catch errors? I wrote it like this initially because I forgot that errors will propagate to a higher level until caught or the program crashes. I was trying to ensure that calling this function inside a `try...catch` block would have the error thrown up to the the calling bloc but it already does that because of error propagation lol. :man_facepalming:

---

I would always work backwards from what our API needs to look like. In the case of `RaspGpio#read`, since it's an `async/await` API, throwing the error seems appropriate. However in the case of `RaspGpio#subscribe`, the error needs to be passed to a callback instead of being thrown.